cmd/log/src/lib.rs

```diff
@@ -0,0 +1,338 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! ## `humility log`
+//!
+//! `humility log` reads and displays any Hubris string buffers (as created
+//! via the `stringbuf!` macro in the Hubris `ringbuf` crate).  e.g.:
+//!
+//! ```console
+//! % humility log -m LOG_RINGBUF
+//! humility: attached via J-Link
+//! humility: ring buffer task_jefe::LOG_RINGBUF in jefe:
+//! instruction
+//! Task #2 Memory fault at address 0x0
+//! Task #2 Illegal instruction
+//! Task #2 Memory fault at address 0x0
+//! Task #2 Illegal instruction
+//! Task #2 Memory fault at address 0x0
+//! Task #2 Illegal instruction
+//! ...
+//! ```
+//!
+//! Use `-m` or `--monitor` to continuously monitor the buffer, otherwise will just print the
+//! current log and exit.
+//!
+//! If an argument is provided, only string buffers that have a name that
+//! contains the argument as a substring, or are in a task that contains
+//! the argument as a substring will be displayed.  
+//!
+//! See the [`ringbuf`
+//! documentation](https://github.com/oxidecomputer/hubris/blob/master/lib/ringbuf/src/lib.rs) for more details.
+
+use anyhow::{bail, Result};
+use clap::Command as ClapCommand;
+use clap::{CommandFactory, Parser};
+use humility::core::Core;
+use humility::hubris::*;
+use humility::reflect::{self, Load, Value};
+use humility_cmd::doppel::{StaticCell, Stringbuf};
+use humility_cmd::{Archive, Attach, Command, Run, Validate};
+use std::collections::BTreeMap;
+use std::iter::zip;
+use std::num::ParseIntError;
+use std::str;
+use std::thread;
+use std::time;
+
+#[derive(Parser, Debug)]
+#[clap(name = "stringbuf", about = env!("CARGO_PKG_DESCRIPTION"))]
+struct StringbufArgs {
+    /// list variables
+    #[clap(long, short)]
+    list: bool,
+    /// continously update the log
+    #[clap(long, short)]
+    monitor: bool,
+    /// how long to delay between polling for new log entries
+    #[clap(long, short, default_value = "100")]
+    delay: u64,
+    /// print only a single stringbuffer by substring of name
+    #[clap(conflicts_with = "list")]
+    name: Option<String>,
+}
+
+///
+/// Extract the timestamp from the line
+///
+fn extract_timestamp(l: &str) -> Result<u64, ParseIntError> {
+    let mut line = l.split(':');
+    // the first element before the ':' should be the timestamp
+    let hopefully_number_part = line.next().unwrap();
+    hopefully_number_part.parse()
+}
+
+fn stringbuf_monitor(
+    hubris: &HubrisArchive,
+    core: &mut dyn Core,
+    definition: &HubrisStruct,
+    ringbuf_var: &HubrisVariable,
+    delay: Option<u64>,
+    once: bool,
+) -> Result<()> {
+    let mut buf: Vec<u8> = vec![];
+    buf.resize_with(ringbuf_var.size, Default::default);
+
+    let mut update_stringbuf = || -> Stringbuf {
+        let _info = core.op_start().unwrap();
+        core.read_8(ringbuf_var.addr, buf.as_mut_slice()).unwrap();
+        core.op_done().unwrap();
+
+        // There are two possible shapes of ringbufs, depending on the age of the
```

Comment:
I think this comment applies only to Ringbuf and not to Stringbuf. Looks like Ringbuf had a different format previously and now they use a different format. However, that doesn't apply to Stringbuf.

Reply:
Yea, that was an outdated comment, the code below it only checked for the one stringbuf type.